### PR TITLE
containerd integration: compute virtualsize

### PR DESCRIPTION
**Pull request by thaJeztah**
- original PR https://github.com/rumpl/moby/pull/6

Upstreaming more commits from https://github.com/rumpl/moby

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/containerd/service.go [hunk 1]
```diff
@@ -74,19 +76,25 @@ func (cs *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 	}
 
 	var ret []*types.ImageSummary
+	snapshotter := cs.client.SnapshotService(containerd.DefaultSnapshotter)
```
Comment (AkihiroSuda) [inline daemon/containerd/service.go:79]: > DefaultSnapshotter

Will this be configurable in a follow-up PR?

LGTM then

Comment (thaJeztah) [reply daemon/containerd/service.go:79]: Yes, we should make this configurable; looks like https://github.com/rumpl/moby is currently only using the default (but that's hard-coded in containerd to `overlayfs` for Linux), and we should make that configurable, similar to the existing graph drivers (and have feature detection, likely, if we also want automatic selection)

./cc @rumpl

Comment (rumpl) [reply daemon/containerd/service.go:79]: Indeed we haven't yet added this but we definitely need it, there are some docker commands that will need to have a `--snapshotter` flag.

File: daemon/containerd/image_list.go [hunk 2]
```diff
@@ -35,3 +45,19 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
...
 }
+
+func computeVirtualSize(ctx context.Context, image containerd.Image, snapshotter snapshots.Snapshotter) (int64, error) {
+	var virtualSize int64
+	diffIDs, err := image.RootFS(ctx)
+	if err != nil {
+		return virtualSize, err
+	}
+	for _, chainID := range identity.ChainIDs(diffIDs) {
+		usage, err := snapshotter.Usage(ctx, chainID.String())
```
Comment (thaJeztah) [inline daemon/containerd/image_list.go:56]: FWIW; some optimisations coming when upstreaming https://github.com/rumpl/moby/pull/17 - let me have a look if I can include that one in this PR

Comment (thaJeztah) [reply daemon/containerd/image_list.go:56]: Actually, that makes things a bit more complicated; let me keep it separate for now

**Review by corhere — APPROVED** [review]

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Pull request merged**